### DISCUSSION TITLE
TemplatedControl's resources not populated in App.Resources nor Styles.Resources

### DISCUSSION THREAD

**Discussion by rullinoiz** [AvaloniaUI/Avalonia#15483] [Q&A]
I'm trying to set up a Resources.axaml file for colors being shared across my project, and in one particular TemplatedControl (nothing else), those resources will not show up at all.  This issue only happens with ResourceIncludes in the App.Resources and in Styles.Resources (inside of said TemplatedControl). The other weird thing is that nothing is being seen at all inside of the Styles.Resources block in that particular TemplatedControl. I even put breakpoints and checked Resources, Application.Current.Resources, and neither of them had the resources I needed.

What I believed to be happening is there are different controls in different windows (this is a multiwindow application), but that doesn't explain why the controls in the second window can see the fonts just fine even after moving them to a separate Resources.axaml file. I'm seriously stumped here

the entire project in particular: https://github.com/rullinoiz/avaloniaui-scorebug/tree/avalonia-issue-1 (branch avalonia-issue-1)

the files:
BasePickerControl.axaml (the culprit) https://github.com/rullinoiz/avaloniaui-scorebug/blob/avalonia-issue-1/scoreboard2/scoreboard2/Controls/BasePickerControl.axaml
```xml
<Styles xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:c="using:scoreboard2.Controls">
    <Design.PreviewWith>
        <c:BasePickerControl />
    </Design.PreviewWith>
    
    <Style Selector="c|BasePickerControl /template/ :is(Button)">
        <Setter Property="Background" Value="{StaticResource OffColorBrush}" />
        <Setter Property="Width" Value="20" />
        <Setter Property="Height" Value="20" />
        <Setter Property="Command" 
                Value="{Binding Base.ToggleKey, 
                RelativeSource={RelativeSource TemplatedParent}}" />
        <Setter Property="CommandParameter" Value="{Binding $self.Name}" />
    </Style>
    
    <Styles.Resources>
        <ResourceDictionary>
            <ResourceDictionary.MergedDictionaries>
                <ResourceInclude Source="Common/Resources.axaml" />
            </ResourceDictionary.MergedDictionaries>
        </ResourceDictionary>
    </Styles.Resources>

    <Style Selector="c|BasePickerControl">
        <!-- Set Defaults -->
        <Setter Property="Template">
            <ControlTemplate>
                <c:ConfigEnclosureControl Title="BASE PICKER">
                    <Border VerticalAlignment="Center" HorizontalAlignment="Center">
                        <Border.RenderTransform>
                            <RotateTransform Angle="45" />
                        </Border.RenderTransform>
                        <StackPanel Orientation="Vertical" Spacing="5" VerticalAlignment="Center" HorizontalAlignment="Center">
                            <StackPanel Orientation="Horizontal" VerticalAlignment="Center" Spacing="5">
                                <Button Name="Base1" />
                                <Button Name="Base2" />
                            </StackPanel>
                            <Button HorizontalAlignment="Left" Name="Base3" />
                        </StackPanel>
                    </Border>
                </c:ConfigEnclosureControl>
            </ControlTemplate>
        </Setter>
    </Style>
</Styles>

```

BasePickerControl.axaml.cs https://github.com/rullinoiz/avaloniaui-scorebug/blob/avalonia-issue-1/scoreboard2/scoreboard2/Controls/BasePickerControl.axaml.cs
```c#
using System;
using Avalonia;
using Avalonia.Controls;
using Avalonia.Controls.Primitives;
using Avalonia.Media;
using scoreboard2.Common;
using scoreboard2.Models.Baseball;

namespace scoreboard2.Controls;

public class BasePickerControl : TemplatedControl
{
    public static readonly StyledProperty<Base?> BaseProperty = AvaloniaProperty.Register<BasePickerControl, Base?>(
        "Base");

    public Base? Base
    {
        get => GetValue(BaseProperty);
        set => SetValue(BaseProperty, value);
    }

    private static IBrush? _onColor;
    private static IBrush? _offColor;

    private Button? _base1;
    private Button? _base2;
    private Button? _base3;

    public BasePickerControl()
    {
        var x = Application.Current;
        
        // this.GetResource<>() wraps this.TryFindResource for less typing (/Common/Helper.cs)
        _onColor ??= this.GetResource<IBrush>("OnColor");
        _offColor ??= this.GetResource<IBrush>("OffColor");
    }

    protected override void OnApplyTemplate(TemplateAppliedEventArgs e)
    {
        base.OnApplyTemplate(e);
        
        Console.WriteLine(_onColor is null);
        

        Base ??= new Base();

        _base1 = e.NameScope.Find<Button>(name: "Base1");
        _base2 = e.NameScope.Find<Button>(name: "Base2");
        _base3 = e.NameScope.Find<Button>(name: "Base3");

        Base.PropertyChanged += (_, _) =>
        {
            // if (args.PropertyName != nameof(Base.BaseStates)) return;
            _base1!.Background = Base.BaseStates[0] ? _onColor : _offColor;
            _base2!.Background = Base.BaseStates[1] ? _onColor : _offColor;
            _base3!.Background = Base.BaseStates[2] ? _onColor : _offColor;
        };
    }
}
```

Controls/Common/Resources.axaml (the file that won't work with BasePickerControl.axaml) https://github.com/rullinoiz/avaloniaui-scorebug/blob/avalonia-issue-1/scoreboard2/scoreboard2/Controls/Common/Resources.axaml
```xml
<ResourceDictionary xmlns="https://github.com/avaloniaui"
                    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml">
    
    <SolidColorBrush x:Key="OnColorBrush" Color="#ffffa709" />
    <SolidColorBrush x:Key="OffColorBrush" Color="#ff442d0a" />
    
</ResourceDictionary>
```

**Comment by maxkatz6** [comment]
Haven't debugged it, but BasePickerControl constructor will be executed before control is attached to the tree. Can you probably move it to the OnApplyTemplate?

**Reply by rullinoiz** [reply]
unfortunately still doesn't work, the resource list is completely empty at any given point
<img width="546" alt="Screenshot 2024-04-23 at 23 25 04" src="https://github.com/AvaloniaUI/Avalonia/assets/46585763/08b89456-7253-4d65-a229-3707fe5e9c67">

**Reply by maxkatz6** [reply]
Resource list is expected to be empty, if you didn't put anything directly to this control's resources list.
(Try)FindResource extension specifically iterates through visual tree from current control to the root (and Application), looking for the resource.

**Reply by rullinoiz** [reply]
`this.GetResource<>()` wraps TryFindResource (in Common/Helper.cs) and still doesn't work

**Reply by timunie** [reply]
Can you make this a very minimal sample please? We have no capacity to debug entire apps but community can help with minor things often.
